build/S3Validator/Program.cs

```diff
@@ -0,0 +1,133 @@
+﻿// Copyright 2020 New Relic, Inc. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+using CommandLine;
+
+namespace S3Validator
+{
+    internal class Program
+    {
+        private const string VersionToken = @"{version}";
+
+        private static readonly HttpClient _client = new();
+
+        static void Main(string[] args)
+        {
+            var result = Parser.Default.ParseArguments<Options>(args)
+                .WithParsed(ValidateOptions)
+                .WithNotParsed(HandleParseError);
+
+            var version = result.Value.Version;
+            var configuration = LoadConfiguration(result.Value.ConfigurationPath);
+            var isValid = Validate(version, configuration);
+            if (!isValid)
+            {
+                ExitWithError(ExitCode.Error, "Validation failed for an unknown reason.");
+            }
+
+            Console.WriteLine("Valid.");
+        }
+
+        private static bool Validate(string version, Configuration configuration)
+        {
+            var tasks = new List<Task<HttpResponseMessage>>();
+            foreach (var dir in configuration.DirectoryList!)
+            {
+                foreach (var fileDetail in configuration.FileList!)
+                {
+                    var url = $"{configuration.BaseUrl}/{dir.Replace(VersionToken, version)}/{fileDetail.Name.Replace(VersionToken, version)}";
+                    var request = new HttpRequestMessage(HttpMethod.Head, url);
+                    request.Options.TryAdd("expectedSize", fileDetail.Size);
+                    tasks.Add(_client.SendAsync(request));
+                }
+            }
+
+            if (!tasks.Any())
+            {
+                return false;
+            }
+
+            var taskCompleted = Task.WaitAll(tasks.ToArray(), 10000);
+            if (!taskCompleted)
+            {
+                ExitWithError(ExitCode.Error, "Validation timed out waiting for HttpClient requests to complete.");
+            }
+
+            var isValid = true;
+            var results = new List<string>();
+            foreach (var task in tasks)
+            {
+                var status = "Valid";
+
+                if (!task.Result.IsSuccessStatusCode)
+                {
+                    isValid = false;
+                    status = task.Result.StatusCode.ToString();
+                }
+                else if (task.Result.Content.Headers.ContentLength < task.Result.RequestMessage?.Options.GetValue<long>("expectedSize"))
+                {
+                    isValid = false;
+                    status = "FileSize";
+                }
+
+                results.Add($"{status,-12}{task.Result.RequestMessage?.RequestUri}");
+            }
+
+            if (!isValid)
+            {
+                ExitWithError(ExitCode.Error, "Validation failed. Results:" + Environment.NewLine + string.Join(Environment.NewLine, results));
+            }
+
+            return isValid;
```

Comment:
Do we need to return isValid here? It seems like the pattern being used is to log errors and exit immediately, so the method would always return true.

Reply:
Updated the logic and removed the need for a return.